README.md
stochastic
=====

Version 0.1.0

Haskell library for monadic composition of probabilistic functions 
and construction of stochastic processes.

Resulting computations can then be run to generate data.

Blog [post](http://kevinl.io/posts/2016-08-17-sampling-monad.html) has some along with examples.

TODO
-----
- See where performance can be improved.
- Check if compiling with -O2 really gives a performance improvement.
- Upload charts with profiled tests and memory usage.
- Add more usage information to README.
- Implement learning for generative models.
- Extend StochProcess to allow for ARCH models.
- Separate StochProcess and Sample modules.
- See if building on top of mwc-random is better than
carrying around StdGens.
- Reduce composed probability distributions that have closed-form
solutions before sampling.


ChangeLog.md
# Revision history for stat-sampling

## 0.1.0.0  -- 2016-08-19

* First version. Released on an unsuspecting world.

## 0.1.1.0  -- 2016-08-26

* Exposes composeProcess for quickly composing StochProcesses.
* Un-reinvents the State monad.
* Adds gamma and beta distributions.


test/Beta.hs
import Control.Monad.Trans

import Data.Foldable
import Data.Stochastic
import Data.Stochastic.Types
import Data.Stochastic.Internal
import Data.Stochastic.Chart

import qualified Data.Sequence as S

import System.Random
import Graphics.Rendering.Chart.Backend.Cairo
import Graphics.Rendering.Chart.Easy

beta1 :: Sampler Double
beta1 = beta 0.5 0.5

beta2 :: Sampler Double
beta2 = beta 0.5 2

gamma1 :: Sampler Double
gamma1 = gamma 3 2

gamma2 :: Sampler Double
gamma2 = gamma 3 0.5

main = do
    gen <- newStdGen
    let betas1 = sampleN 100000 beta1 gen
    let betas2 = sampleN 100000 beta2 gen
    let gammas1 = sampleN 100000 gamma1 gen
    let gammas2 = sampleN 100000 gamma2 gen
    renderableToFile def "charts/beta1.png" $ toRenderable $ histogram "Beta1" $ toList betas1
    renderableToFile def "charts/beta2.png" $ toRenderable $ histogram "Beta2" $ toList betas2
    renderableToFile def "charts/gamma1.png" $ toRenderable $ histogram "Gamma1" $ toList gammas1
    renderableToFile def "charts/gamma2.png" $ toRenderable $ histogram "Gamma2" $ toList gammas2


test/Test.hs
module Main where

import Data.Stochastic
import Data.Stochastic.Internal
import Data.Stochastic.Types

import qualified Data.Sequence as S

fairCoin :: Sampler Bool
fairCoin = bernoulli 0.5

shadyGambler :: Bool -> Sampler Double
shadyGambler b = if b then discrete [(100, 0.9), (-100, 0.1)]
                 else certain (-100)

gamblerGivenCoin :: Sampler Bool -> Sampler Double
gamblerGivenCoin coin = coin >>= shadyGambler

main = do
    s <- sampleION 1000000 $ gamblerGivenCoin fairCoin
    let wins = S.filter (== 100) s
        losses = S.filter (== (-100)) s
    print $ "Wins with fair coin: " ++ (show $ length wins)
    print $ "Losses with fair coin: " ++ (show $ length losses)


test/Normal3.hs
import Data.Stochastic
import Data.Stochastic.Types
import Data.Stochastic.Internal

import System.Random

normalInit :: StochProcess
normalInit = normalProcess 10 3

f :: Double -> StochProcess
f d = if d < 7 then normalProcess (0.9 * d) (0.3 * d)
  else if d > 13 then normalProcess (1.1 * d) (0.3 * d)
  else normalProcess d (0.3 * d)

normal3 :: StochProcess
normal3 = composeProcess 3 normalInit f

main = do
    gen <- newStdGen
    print "normal3"
    print $ runProcessN 100000 normal3 gen


src/Data/Sample.hs
{-|
 Module         : Data.Sample
 Description    : Types used for Sample.
 License        : GPL-3
 Maintainer     : hackage@mail.kevinl.io
 Stability      : experimental

 This module contains convenience functions to
 construct 'Sample's or 'StochProcess'es corresponding
 to several probability distributions.

 It also contains functions that can be used for
 running the constructed 'StochProcess'es and generating
 datapoints, or sampling from a constructed 'Sample'.

 Some examples for usage can be found here: <http://kevinl.io/posts/2016-08-17-sampling-monad.html>
-}

module Data.Sample where

import Control.Monad.Trans
import Control.Monad.Writer

import Data.Sample.Types
import qualified Data.Sequence as S

import System.Random

-- | Function to construct a 'StochProcess' computation
-- given an initial computation, a 'StochProcess' function,
-- and number of times to apply the function with bind.
composeProcess :: Integral i => i -> StochProcess -> (Double -> StochProcess) -> StochProcess
composeProcess i pr f = if i <= 0 then pr 
                   else (composeProcess (i-1) pr f) >>= f

-- | Sample from the 'StochProcess' computation, discarding
-- the new 'RandomGen'.
sampleProcess_ :: StochProcess -> StdGen -> Double
sampleProcess_ ma g = flip sample_ g $ liftM fst $ runWriterT ma

-- | Sample from the 'StochProcess' computation, returning
-- the value of type a and a new 'RandomGen'.
sampleProcess :: StochProcess -> StdGen -> (Double, StdGen)
sampleProcess ma g = flip sample g $ liftM fst $ runWriterT ma

-- | Get a certain number of samples from the 'StochProcess' computation.
sampleProcessN :: (Integral i) => i -> StochProcess -> StdGen -> S.Seq Double
sampleProcessN i ma g = if i <= 0 then S.empty
                   else let (a, gen) = sampleProcess ma g
                   in a S.<| sampleProcessN (i-1) ma gen

-- | Run a 'StochProcess' computation and retrieve the recorded
-- results along with a new 'RandomGen'.
runProcess :: StochProcess -> StdGen -> (S.Seq Double, StdGen)
runProcess ma g = flip sample g $ execWriterT ma

-- | Run a 'StochProcess' computation and retrieve the recorded
-- results, discarding the new 'RandomGen'.
runProcess_ :: StochProcess -> StdGen -> S.Seq Double
runProcess_ ma g = fst $ runProcess ma g

-- | Runs a 'StochProcess' computation a given number times
-- and produces a 'Sequence' of 'Sequence's of Doubles.
-- | Get a certain number of samples from the 'Sample'
runProcessN :: (Integral i) => i -> StochProcess -> StdGen -> S.Seq (S.Seq Double)
runProcessN n pr gen = if n <= 0 then S.empty
                  else let (seq, gen') = runProcess pr gen
                       in seq S.<| runProcessN (n-1) pr gen'

-- | 'StochProcess' sample for a normal distribution that records
-- the value sampled from the normal distribution.
normalProcess :: Mean -> StDev -> StochProcess 
normalProcess mean std = do
    sample <- lift $ normal mean std
    tell $ S.singleton sample
    return sample

-- | 'StochProcess' sample for a distribution over 'Double's that always
-- returns the same value when sampled, and records that value.
certainProcess :: Double -> StochProcess 
certainProcess a = do
    sample <- lift $ certain a
    tell $ S.singleton sample
    return sample

-- | 'StochProcess' sample for a discrete distribution over 'Double's
-- that records the value sampled from the normal distribution.
discreteProcess :: [(Double, Double)] -> StochProcess 
discreteProcess a = do
    sample <- lift $ discrete a
    tell $ S.singleton sample
    return sample

-- | 'StochProcess' sample for a uniform distribution over 'Double's
-- that records the value sampled from it.
uniformProcess :: [Double] -> StochProcess
uniformProcess l = do
    sample <- lift $ uniform l
    tell $ S.singleton sample
    return sample

-- | Function to make a 'Sample' out of a provided
-- 'Distribution'.
mkSample :: (RandomGen g, Sampleable d) => d a -> Sample g d a
mkSample d = Sample $ \g -> (d, snd $ next g)

-- | 'Sample' for a normal distribution with given
-- 'StdGen', 'Mean', and 'StDev'.
normal :: RandomGen g => Mean -> StDev -> Sample g Distribution Double
normal m s = mkSample $ Normal m $ abs s

-- | 'Sample' for a Bernoulli distribution with given
-- probability to produce True.
bernoulli :: RandomGen g => Double -> Sample g Distribution Bool
bernoulli f = mkSample $ Bernoulli f

-- | 'Sample' for a discrete distribution with given
-- list of tuples of values of type a and 'Double's 
-- representing the probability of producing each
-- value when sampling from this distribution.
discrete :: RandomGen g => [(a, Double)] -> Sample g Distribution a
discrete [] = error "do not construct empty discrete distributions"
discrete l = mkSample $ Discrete l

-- | 'Sample' for a uniform distribution
-- given a list of provided values.
uniform :: (RandomGen g) => [a] -> Sample g Distribution a
uniform l = mkSample $ Uniform l

-- | 'Sample' for a distribution where we always sample
-- the same value.
certain :: (RandomGen g, Sampleable d) => a -> Sample g d a
certain = mkSample . certainDist

-- | Get one sample of type a from the 'Sample' along with
-- a new 'StdGen'.
--
-- We do an extra 'next' in order to get one more
-- 'RandomGen' because when we sample from normal
-- distributions, we consume one extra 'RandomGen'.
sample :: (RandomGen g, Sampleable d) => Sample g d a -> g -> (a, g)
sample s g = let (dist, g') = runSample s g
                 (a, g'') = sampleFrom dist g'
             in (a, snd $ next g'')

-- | Get one sample of type a from the 'Sample',
-- discarding the 'RandomGen'
sample_ :: (RandomGen g, Sampleable d) => Sample g d a -> g -> a
sample_ s g = fst $ sample s g

-- | Get a certain number of samples from the 'Sample'
sampleN :: (RandomGen g, Sampleable d, Integral i) => i -> Sample g d a -> g -> S.Seq a
sampleN i s g = if i <= 0 then S.empty
                else let (a, g') = sample s g
                     in a S.<| sampleN (i - 1) s g'

-- | Sample from a 'Sample' of type a using the global
-- random number generator provided by 'newStdGen',
-- returning a new 'StdGen' with the sampled value.
sampleIO :: Sampleable d => Sample StdGen d a -> IO (a, StdGen)
sampleIO s = sample s <$> newStdGen

-- | Sample from a 'Sample' of type a using the global
-- random number generator provided by 'newStdGen',
-- discarding the new 'StdGen'.
sampleIO_ :: Sampleable d => Sample StdGen d a -> IO a
sampleIO_ s = fst <$> (sample s <$> newStdGen)

-- | Produce several samples from the 'Sample' using the random number generator
-- in the IO monad.
sampleION :: (Sampleable d, Integral i) => i -> Sample StdGen d a -> IO (S.Seq a)
sampleION i s = sampleN i s <$> newStdGen

-- | TODO: Concurrent sampling in the IO monad, with 'RandomGen' splitting.


src/Data/Sample/Lib.hs
{-|
 Module         : Data.Sample.Lib
 Description    : Utility functions for Sample types.
 License        : GPL-3
 Maintainer     : hackage@mail.kevinl.io
 Stability      : experimental

-}
module Data.Sample.Lib where

import System.Random

import Numeric.MathFunctions.Constants

-- | Function to convert values sampled from uniform distribution
-- to a value sampled from a standard normal distribution.
boxMuller :: (Floating a) => a -> a -> a
boxMuller u1 u2 = cos (2 * pi * u2) * (sqrt $ (-2) * (log u1))

-- | Randoms that aren't too small or equal to 1.
decentRandom :: (RandomGen g) => g -> (Double, g)
decentRandom gen = let (sampled, newG) = randomR (0, 1.0) gen
                   in if sampled <= m_epsilon || sampled == 1
                      then decentRandom newG
                      else (sampled, newG)


src/Data/Stochastic/Chart.hs
{-|
 Module         : Data.Stochastic.Chart
 Description    : Quick and dirty plotting functions.
 License        : GPL-3
 Maintainer     : hackage@mail.kevinl.io
 Stability      : experimental

 This module contains some very quick and dirty
 functions to plot the results obtained from
 running stochastic processes.

-}
module Data.Stochastic.Chart where

import Control.Monad

import Data.Foldable

import Data.Stochastic.Types

import qualified Data.Sequence as S

import Graphics.Rendering.Chart.Easy
import Graphics.Rendering.Chart.Backend.Cairo

-- | Create lines with the same title.
seqLines :: String -> S.Seq (S.Seq Double) -> EC l (PlotLines Integer Double)
seqLines t ss = line t $ toList $ fmap (\x -> zip [1..] $ toList x) ss

-- | Easy function to make a plot. Really awful, don't use this.
processToChart :: String  -- ^ Title of the plot.
          -> S.Seq (S.Seq Double)  -- ^ Lines to plot.
          -> EC (Layout Integer Double) ()
processToChart t ss = do
    layout_title .= t
    plot $ seqLines t ss

-- | Convenience function to make a histogram.
histogram :: String -> [Double] -> Layout Double Double
histogram title values = layout
    where hist = plot_hist_values  .~ values
                 $ plot_hist_drop_lines .~ True
                 $ defaultFloatPlotHist
          layout :: Layout Double Double
          layout = layout_title .~ title
                 $ layout_plots .~ [ histToPlot hist ]
                 $ def


src/Data/Stochastic/Types.hs
{-#LANGUAGE GADTs#-}
{-#LANGUAGE RankNTypes#-}
{-#LANGUAGE FlexibleInstances#-}

{-|
 Module         : Data.Stochastic.Types
 Description    : Types used for the stochastic package.
 License        : GPL-3
 Maintainer     : hackage@mail.kevinl.io
 Stability      : experimental

 This module contains the types used
 for the stochastic package.

 WARNING: In its current state, care should be
 taken when using discrete distributions
 as it is never checked that the probabilities
 sum to 1. As is, execution of sampling may fail at run-time
 if probabilities aren't normalized.
-}

module Data.Stochastic.Types (
  Distribution (..)
, Sampleable (..)
, Sample (..)
, StochProcess (..)
, Sampler (..)
, Mean (..)
, StDev (..)
, marsagliaTsang
) where

import Control.Monad
import Control.Monad.State
import Control.Monad.Writer

import Data.Stochastic.Internal

import qualified Data.Sequence as S

import System.Random

-- | Datatype representing parameterized probability distributions
-- over values of type a. GADTs are used to restrict types
-- of certain distributions (e.g. normal distributions can
-- only be defined over floating point numbers)
data Distribution a where
    Normal :: Mean -> StDev -> Distribution Double
    Bernoulli :: Double -> Distribution Bool
    Discrete :: [(a, Double)] -> Distribution a
    DiscreteUniform :: [a] -> Distribution a
    Uniform :: Distribution Double
    Certain :: a -> Distribution a
    Gamma :: Double -> Double -> Distribution Double
    -- ^ Gamma distribution, where the first parameter is the
    -- shape parameter alpha, and the second parameter is the
    -- scale parameter beta.
    Beta :: Double -> Double -> Distribution Double

-- | Class of types from which samples can be obtained.
class Sampleable d where
    -- | Constructor for a datatype from which we always
    -- sample the same value.
    certainDist :: a -> d a
    -- | Sample from the sampleable datatype using a 'RandomGen'
    -- returning a new 'RandomGen'.
    sampleFrom :: (RandomGen g) => d a -> g -> (a, g)

-- | 'Sampleable' instance for 'Distribution'. We ensure
-- that we always pass the *next* 'RandomGen' provided
-- to sampleFrom. This lets us obey the monad laws.
instance Sampleable Distribution where
    sampleFrom da g
        = case da of
            Normal mean stdev 
                -> let (a, g')   = closedRnd g 
                       (a', g'') = closedRnd g'
                       s = (stdev * (boxMuller a a')) + mean
                   in (s, g')
            Bernoulli prob    
                -> let (a, g') = closedRnd g
                   in (a <= prob, g')
            Discrete []
                -> error "cannot sample from empty discrete distribution"
            Discrete l        
                -> let (a, g') = closedRnd g
                   in (scan a l, g')
                   where scan lim [] = 
                             if lim <= 0 then error $ "not normalized discrete dist"
                             else error "empty discrete dist"
                         scan lim (x:xs) = 
                             if lim <= snd x then fst x 
                             else scan (lim - snd x) xs
            DiscreteUniform []
                -> error "cannot sample from empty discrete distribution"
            DiscreteUniform l
                -> let (a, g') = closedRnd g
                       prob = 1 / (fromIntegral $ length l)
                   in (l !! (floor $ a / prob), g')
            Uniform
                -> closedRnd g
            Gamma alpha beta
                -> if alpha <= 0 || beta <= 0 then error "alpha and beta parameter cannot be less than or equal to zero in beta distribution"
                   else if alpha > 0 && alpha < 1 then
                        let (a, g') = sampleFrom (Gamma (alpha + 1) beta) g
                            (uni, g'') = openRnd g'
                        in (a * (uni ** (1/alpha)), g'')
                   else let d = alpha - (1/3)
                            c = 1 / sqrt (9 * d)
                            (m, g') =  marsagliaTsang d c g
                        in (m * beta, g')
            Beta alpha beta
                -> let (x, g') = sampleFrom (Gamma alpha 1) g
                       (y, g'') = sampleFrom (Gamma beta 1) g'
                   in (x / (x + y), g'')
            Certain val       
                -> (val, snd $ openRnd g) 
                -- Seemingly unnecessary, but important to obey the monad laws to always produce the same RandomGen each time we sample.
    certainDist = Certain

-- | Marsaglia and Tsang's rejection method
-- for generating Gamma variates with parameters
-- alpha and 1, where 1 is the scale parameter,
-- given d and c.
marsagliaTsang :: (RandomGen g) => Double -> Double -> g -> (Double, g)
marsagliaTsang d c g =
    let (norm, g') = sampleFrom (Normal 0 1) g
        (uni, g'') = sampleFrom Uniform g'
        v = (1 + (c * norm)) ** 3
    in if norm > ((-1)/c) && 
          log uni < ((norm ** 2)/2 + d - (d * v) + (d * log v))
       then (d * v, g'')
       else marsagliaTsang d c g''

-- | Show instance for 'Distribution's.
instance (Show a) => Show (Distribution a) where
    show da = case da of
        Normal mean stdev -> "Normal " ++ show mean ++ " " ++ show stdev
        Bernoulli prob -> "Bernoulli " ++ show prob
        Discrete l -> "Discrete " ++ show l
        DiscreteUniform l -> "DiscreteUniform " ++ show l       
        Certain val -> "Certain " ++ show val

-- | 'Sample' monad containing a random number generator plus a type from which
-- we can sample values of type a
newtype Sample g d a
    = Sample { runSample :: (RandomGen g, Sampleable d) => State g (d a) }

-- | Monad that represents a stochastic process.
-- It allows us to record numeric values as we sample.
type StochProcess
    = WriterT (S.Seq Double) (Sample StdGen Distribution) Double

-- | Monad instance for Sample.
instance (RandomGen g, Sampleable d) => Monad (Sample g d) where
    return x = Sample $ do
                modify (snd . next)
                return $ certainDist x

    (>>=) ma f = Sample $ do
                     modify (snd . next)
                     dist <- runSample ma
                     g <- get
                     let a = fst $ sampleFrom dist g
                     runSample (f a)

-- | Trivial 'Functor' instance for 'Sample' 'StdGen' 'Distribution'.
instance (RandomGen g, Sampleable s) => Functor (Sample g s) where
    fmap = liftM

-- | Trivial 'Applicative' instance for 'Sample' 'StdGen' 'Distribution'.
instance (RandomGen g, Sampleable s) => Applicative (Sample g s) where
    pure = return
    (<*>) = ap

-- | Type synonym for shorter type annotations for 'Sample'.
type Sampler a = Sample StdGen Distribution a

-- | Type synonym for 'Double' so that the 
-- type annotation for the 'Normal' constructor is more informative.
type Mean = Double

-- | Type synonyms for 'Double' so that the 
-- type annotation for the 'Normal' constructor is more informative.
type StDev = Double


src/Data/Stochastic/Internal.hs
{-|
 Module         : Data.Stochastic.Internal
 Description    : Internal utility functions for Sample types.
 License        : GPL-3
 Maintainer     : hackage@mail.kevinl.io
 Stability      : experimental

-}
module Data.Stochastic.Internal (
  boxMuller
, closedRnd
, openRnd
, closedOpenRnd
, openClosedRnd
) where

import System.Random

-- | Function to convert values sampled from uniform distribution
-- to a value sampled from a standard normal distribution.
boxMuller :: (Floating a) => a -> a -> a
boxMuller u1 u2 = cos (2 * pi * u2) * (sqrt $ (-2) * (log u1))

-- | Randoms in the interval [0, 1]
closedRnd :: (RandomGen g) => g -> (Double, g)
closedRnd gen = randomR (0, 1.0) gen

-- | Randoms in the interval (0, 1)
openRnd :: (RandomGen g) => g -> (Double, g)
openRnd gen = let (a, g) = closedRnd gen
                   in if a == 0 || a == 1
                      then openRnd g
                      else (a, g)
                      
-- | Randoms in the interval [0, 1)
closedOpenRnd :: (RandomGen g) => g -> (Double, g)
closedOpenRnd gen = let (a, g) = closedRnd gen
                    in if a == 1
                       then closedOpenRnd g
                       else (a, g)
                      
-- | Randoms in the interval (0, 1]
openClosedRnd :: (RandomGen g) => g -> (Double, g)
openClosedRnd gen = let (a, g) = closedRnd gen
                   in if a == 0
                      then openClosedRnd g
                      else (a, g)


src/Data/Stochastic.hs
{-|
 Module         : Data.Stochastic
 Description    : Main module containing functions for interacting with Samples and StochProcesses.
 License        : GPL-3
 Maintainer     : hackage@mail.kevinl.io
 Stability      : experimental

 This module contains convenience functions to
 construct 'Sample's or 'StochProcess'es corresponding
 to several probability distributions.

 It also contains functions that can be used for
 running the constructed 'StochProcess'es and generating
 datapoints, or sampling from a constructed 'Sample'.

 Some examples for usage can be found here: <http://kevinl.io/posts/2016-08-17-sampling-monad.html>
-}

module Data.Stochastic (
  -- * Constructing a Sample
  mkSample
, certain
, discreteUniform
, discrete
, bernoulli
, normal
, uniform
, gamma
, beta
  -- * Sampling from a Sample
, sample
, sample_
, sampleN
, sampleIO
, sampleIO_
, sampleION
  -- * Constructing a StochProcess
, toProcess
, certainProcess
, discreteUniformProcess
, discreteProcess
, normalProcess
, uniformProcess
, gammaProcess
, betaProcess
, composeProcess
  -- * Running a StochProcess
, runProcess
, runProcess_
, runProcessN
  -- * Sampling from a StochProcess
, sampleProcess
, sampleProcess_
, sampleProcessN
, sampleProcessIO
, sampleProcessION
  -- * The types
, module Data.Stochastic.Types
  -- * Internal functions for your viewing pleasure
, module Data.Stochastic.Internal
) where

import Control.Monad.State
import Control.Monad.Trans
import Control.Monad.Writer

import Data.Stochastic.Internal
import Data.Stochastic.Types

import qualified Data.Sequence as S

import System.Random

-- | Function to construct a 'StochProcess' computation
-- given an initial computation, a 'StochProcess' function,
-- and number of times to apply the function with bind.
composeProcess :: Integral i => i -> StochProcess -> (Double -> StochProcess) -> StochProcess
composeProcess i pr f = if i <= 0 then pr 
                   else (composeProcess (i-1) pr f) >>= f

-- | Sample from the 'StochProcess' computation, discarding
-- the new 'RandomGen'.
sampleProcess_ :: StochProcess -> StdGen -> Double
sampleProcess_ ma g = flip sample_ g $ liftM fst $ runWriterT ma

-- | Sample from the 'StochProcess' computation, returning
-- the value of type a and a new 'RandomGen'.
sampleProcess :: StochProcess -> StdGen -> (Double, StdGen)
sampleProcess ma g = flip sample g $ liftM fst $ runWriterT ma

-- | Sample from the 'StochProcess' computation
-- in the IO monad, returning a 'Double'
-- and a 'RandomGen' created in the IO monad.
sampleProcessIO :: StochProcess -> IO (Double, StdGen)
sampleProcessIO ma = do
    gen <- newStdGen
    return $ sampleProcess ma gen

-- | Sample from the 'StochProcess' computation
-- in the IO monad, returning a 'Double'
-- and discarding the 'RandomGen'.
sampleProcessIO_ :: StochProcess -> IO Double
sampleProcessIO_ ma = fst <$> sampleProcessIO ma

-- | Get a certain number of samples from the 'StochProcess' computation in the IO monad.
sampleProcessION :: (Integral i) => i -> StochProcess -> IO (S.Seq Double)
sampleProcessION i ma = do
    gen <- newStdGen
    return $ sampleProcessN i ma gen

-- | Get a certain number of samples from the 'StochProcess' computation.
sampleProcessN :: (Integral i) => i -> StochProcess -> StdGen -> S.Seq Double
sampleProcessN i ma g = if i <= 0 then S.empty
                   else let (a, gen) = sampleProcess ma g
                   in a S.<| sampleProcessN (i-1) ma gen

-- | Run a 'StochProcess' computation and retrieve the recorded
-- results along with a new 'RandomGen'.
runProcess :: StochProcess -> StdGen -> (S.Seq Double, StdGen)
runProcess ma g = flip sample g $ execWriterT ma

-- | Run a 'StochProcess' computation and retrieve the recorded
-- results, discarding the new 'RandomGen'.
runProcess_ :: StochProcess -> StdGen -> S.Seq Double
runProcess_ ma g = fst $ runProcess ma g

-- | Runs a 'StochProcess' computation a given number times
-- and produces a 'Sequence' of 'Sequence's of Doubles.
runProcessN :: (Integral i) => i -> StochProcess -> StdGen -> S.Seq (S.Seq Double)
runProcessN n pr gen = if n <= 0 then S.empty
                  else let (seq, gen') = runProcess pr gen
                       in seq S.<| runProcessN (n-1) pr gen'

-- | 'StochProcess' sample for a normal distribution that records
-- the value sampled from the normal distribution.
normalProcess :: Mean -> StDev -> StochProcess 
normalProcess mean std = toProcess $ normal mean std

-- | 'StochProcess' for a distribution over 'Double's that always
-- returns the same value when sampled, and records that value.
certainProcess :: Double -> StochProcess 
certainProcess a = toProcess $ certain a

-- | 'StochProcess' for a discrete distribution over 'Double's
-- that records the value sampled from the normal distribution.
discreteProcess :: [(Double, Double)] -> StochProcess 
discreteProcess a = toProcess $ discrete a

-- | 'StochProcess' for a discrete uniform distribution 
-- over 'Double's that records the value sampled from it.
discreteUniformProcess :: [Double] -> StochProcess
discreteUniformProcess l = toProcess $ discreteUniform l

-- | 'StochProcess' for a discrete uniform distribution 
-- over 'Double's that records the value sampled from it.
uniformProcess :: StochProcess
uniformProcess = toProcess uniform

-- | 'StochProcess' for a gamma distribution with
-- provided shape and scale parameters.
gammaProcess :: Double -> Double -> StochProcess
gammaProcess a b = toProcess $ gamma a b

-- | 'StochProcess' for a beta distribution.
betaProcess :: Double -> Double -> StochProcess
betaProcess a b = toProcess $ beta a b

-- | Function to create a 'StochProcess' out of a provided
-- 'Sample' over 'Double's.
toProcess :: Sample StdGen Distribution Double -> StochProcess
toProcess s = do
    sample <- lift s
    tell $ S.singleton sample
    return sample

-- | 'Sample' for a normal distribution with given
-- 'StdGen', 'Mean', and 'StDev'.
normal :: RandomGen g => Mean -> StDev -> Sample g Distribution Double
normal m s = mkSample $ Normal m $ abs s

-- | 'Sample' for a Bernoulli distribution with given
-- probability to produce True.
bernoulli :: RandomGen g => Double -> Sample g Distribution Bool
bernoulli f = mkSample $ Bernoulli f

-- | 'Sample' for a discrete distribution with given
-- list of tuples of values of type a and 'Double's 
-- representing the probability of producing each
-- value when sampling from this distribution.
discrete :: RandomGen g => [(a, Double)] -> Sample g Distribution a
discrete [] = error "do not construct empty discrete distributions"
discrete l = mkSample $ Discrete l

-- | 'Sample' for a discrete uniform distribution
-- given a list of provided values.
discreteUniform :: (RandomGen g) => [a] -> Sample g Distribution a
discreteUniform l = mkSample $ DiscreteUniform l

-- | 'Sample' for a continuous uniform distribution
-- with support [0, 1].
uniform :: (RandomGen g) => Sample g Distribution Double
uniform = mkSample Uniform

-- | 'Sample' for a distribution where we always sample
-- the same value.
certain :: (RandomGen g, Sampleable d) => a -> Sample g d a
certain = mkSample . certainDist

-- | 'Sample' for a gamma distribution given shape parameter
-- and scale parameter.
gamma :: RandomGen g
      => Double 
      -- ^ The shape parameter.
      -> Double 
      -- ^ The scale parameter.
      -> Sample g Distribution Double
gamma a b = if a <= 0 || b <= 0 
            then error "cannot construct gamma dist with <=0 params"
            else mkSample $ Gamma a b

-- | 'Sample' for a beta distribution.
beta :: RandomGen g => Double -> Double -> Sample g Distribution Double
beta a b = if a <= 0 || b <= 0 
            then error "cannot construct gamma dist with <=0 params"
            else mkSample $ Beta a b

-- | Get one sample of type a from the 'Sample' along with
-- a new 'StdGen'.
--
-- We do an extra 'next' in order to get one more
-- 'RandomGen' because when we sample from normal
-- distributions, we consume one extra 'RandomGen'.
sample :: (RandomGen g, Sampleable d) => Sample g d a -> g -> (a, g)
sample s g = let (dist, g') = runState (runSample s) g
                 (a, g'') = sampleFrom dist g'
             in (a, snd $ next g'')

-- | Get one sample of type a from the 'Sample',
-- discarding the 'RandomGen'
sample_ :: (RandomGen g, Sampleable d) => Sample g d a -> g -> a
sample_ s g = fst $ sample s g

-- | Get a certain number of samples from the 'Sample'
sampleN :: (RandomGen g, Sampleable d, Integral i) => i -> Sample g d a -> g -> S.Seq a
sampleN i s g = if i <= 0 then S.empty
                else let (a, g') = sample s g
                     in a S.<| sampleN (i - 1) s g'

-- | Sample from a 'Sample' of type a using the global
-- random number generator provided by 'newStdGen',
-- returning a new 'StdGen' with the sampled value.
sampleIO :: Sampleable d => Sample StdGen d a -> IO (a, StdGen)
sampleIO s = sample s <$> newStdGen

-- | Sample from a 'Sample' of type a using the global
-- random number generator provided by 'newStdGen',
-- discarding the new 'StdGen'.
sampleIO_ :: Sampleable d => Sample StdGen d a -> IO a
sampleIO_ s = fst <$> (sample s <$> newStdGen)

-- | Produce several samples from the 'Sample' using the random number generator
-- in the IO monad.
sampleION :: (Sampleable d, Integral i) => i -> Sample StdGen d a -> IO (S.Seq a)
sampleION i s = sampleN i s <$> newStdGen

-- | Function to make a 'Sample' out of a provided
-- 'Distribution'.
mkSample :: (RandomGen g, Sampleable d) => d a -> Sample g d a
mkSample d = Sample $ return d
